server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -848,17 +927,50 @@ public void deleteStaleSegmentsAsync(int lastNMetadataFilesToKeep, ActionListene
         }
     }
 
+    public static boolean cleanupAsync(
+        RemoteSegmentStoreDirectoryFactory remoteDirectoryFactory,
+        ThreadPool threadpool,
+        String remoteStoreRepoForIndex,
+        String indexUUID,
+        ShardId shardId
+    ) {
+        try {
+            threadpool.executor(ThreadPool.Names.REMOTE_PURGE).execute(() -> {
+                try {
+                    try (
+                        RemoteSegmentStoreDirectory remoteDirectory = (RemoteSegmentStoreDirectory) remoteDirectoryFactory.newDirectory(
+                            remoteStoreRepoForIndex,
+                            indexUUID,
+                            shardId
+                        )
+                    ) {
+                        remoteDirectory.deleteStaleSegments(0);
+                        remoteDirectory.deleteIfEmpty();
+                    }
+                } catch (IOException e) {}
+            });
+        } catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
     /*
     Tries to delete shard level directory if it is empty
     Return true if it deleted it successfully
      */
     private boolean deleteIfEmpty() throws IOException {
-        Collection<String> metadataFiles = remoteMetadataDirectory.listFilesByPrefixInLexicographicOrder(
-            MetadataFilenameUtils.METADATA_PREFIX,
-            1
-        );
-        if (metadataFiles.size() != 0) {
-            logger.info("Remote directory still has files, not deleting the path");
```

Comment:
but since we will be calling this method during or after the index deletion, it should be able to cleanup the md files which were failed to delete earlier. we can consider lock files as a check which will prevent deletion for a md file.

Reply:
since single segment file failure can cause md file to not delete, i think lock files are a better check to prevent deletion for md which are currently in use and anyway we call this method at the time of directory cleanup. but at the same time, i agree that there need to be a separate GC mechanism to handle the failure scenarios. so just reverting this code back for now.